test/ng/rootScopeSpec.js

```diff
@@ -2316,6 +2316,97 @@ describe('Scope', function() {
         }));
 
 
+        // See issue https://github.com/angular/angular.js/issues/16135
+        it('should deallocate the listener array entry', inject(function($rootScope) {
+          var remove1 = $rootScope.$on('abc', noop);
+          $rootScope.$on('abc', noop);
+
+          expect($rootScope.$$listeners['abc'].length).toBe(2);
+
+          remove1();
+
+          expect($rootScope.$$listeners['abc'].length).toBe(1);
+        }));
+
+
+        it('should call next listener when removing current', inject(function($rootScope) {
+          var listener1 = jasmine.createSpy().and.callFake(function() { remove1(); });
+          var remove1 = $rootScope.$on('abc', listener1);
+
+          var listener2 = jasmine.createSpy();
+          var remove2 = $rootScope.$on('abc', listener2);
+
+          var listener3 = jasmine.createSpy();
+          var remove3 = $rootScope.$on('abc', listener3);
+
+          $rootScope.$broadcast('abc');
+          expect(listener1).toHaveBeenCalled();
+          expect(listener2).toHaveBeenCalled();
+          expect(listener3).toHaveBeenCalled();
+
+          listener1.calls.reset();
+          listener2.calls.reset();
+          listener3.calls.reset();
+
+          $rootScope.$broadcast('abc');
+          expect(listener1).not.toHaveBeenCalled();
+          expect(listener2).toHaveBeenCalled();
+          expect(listener3).toHaveBeenCalled();
+        }));
+
+
+        it('should call all listeners when removing previous', inject(function($rootScope) {
+          var listener1 = jasmine.createSpy();
+          var remove1 = $rootScope.$on('abc', listener1);
+
+          var listener2 = jasmine.createSpy().and.callFake(remove1);
+          var remove2 = $rootScope.$on('abc', listener2);
+
+          var listener3 = jasmine.createSpy();
+          var remove3 = $rootScope.$on('abc', listener3);
+
+          $rootScope.$broadcast('abc');
+          expect(listener1).toHaveBeenCalled();
+          expect(listener2).toHaveBeenCalled();
+          expect(listener3).toHaveBeenCalled();
+
+          listener1.calls.reset();
+          listener2.calls.reset();
+          listener3.calls.reset();
+
+          $rootScope.$broadcast('abc');
+          expect(listener1).not.toHaveBeenCalled();
+          expect(listener2).toHaveBeenCalled();
+          expect(listener3).toHaveBeenCalled();
+        }));
+
+
+        it('should not call listener when removed by previous', inject(function($rootScope) {
+          var listener1 = jasmine.createSpy();
+          var remove1 = $rootScope.$on('abc', listener1);
+
+          var listener2 = jasmine.createSpy().and.callFake(function() { remove3(); });
+          var remove2 = $rootScope.$on('abc', listener2);
+
+          var listener3 = jasmine.createSpy();
+          var remove3 = $rootScope.$on('abc', listener3);
```

Comment:
It might be worth adding a 4th listener to make sure the rest of the queue is not messed up.